RFCs/0000-add-pdk-config.md

```diff
@@ -0,0 +1,292 @@
+| user.forge.username | Primary Forge username of this developer. Note that this may be an organization rather than a personal username. Used as the default value in metadata.json when generating a new module. Supersedes previous ~/.pdk/cache/answers.json functionality. |
+| [user.forge.api\_key] | Placeholder for future Forge API interactions. |
+| [user.forge.api\_secret] | Placeholder for future Forge API interactions. |
+
+#### Module-specific Config
+
+The values for these config keys will be persisted in a new .pdk/config file within the module, the new file is intended
+to be checked into VCS and shared with other developers to ensure consistent application of the relevant PDK commands.
+
+| Key | Description |
+| --- | --- |
+| module.template.url | Template URL to use for this module. |
+| module.template.ref | Template ref to use for this module. Note that this will differ from what is currently stored in metadata.json, this value will be the “desired” reference (e.g. a branch name) but not the specific commit that the module was last updated to. |
+| module.template.custom.[...] | Filename based keys with associated config hashes to govern how the template should be applied to this specific module. Supersedes existing .sync.yml settings. |
+| module.test.fixtures.[...] | List of additional modules, etc. and associated config hashes needed in order to run unit tests for this module. Supersedes existing .fixtures.yml settings. |
+| module.x.[...] | Arbitrary keys and values to be utilized by users to persist data between module developers. Can be used as a basic shared data store for custom developer Rake tasks, etc. |
+
+### More About `module.x`
+
+The `module.x` configuration key space is intended to be used by module developers and third-party tool developers to
+persist configuration data that is relevant to all developers of a module. Ultimately we want to avoid any need for
+users to add additional configuration files to their modules beyond what `pdk config` uses.
+
+We will define and document best practices around the usage of the `module.x` key space for the community, including
+things like:
+
+ - Namespace your custom configuration settings under a key that matches the name of your organization or tool name,
+   e.g. `module.x.mycompany`, `module.x.voxpupuli`, or `module.x.super_module_tool`.
+
+ - If your custom configuration is tool-specific, include a key and value that you can use to version your other
+   configuration data. That way you can significantly change the configuration format for your tool in the future
+   without breaking older versions. E.g. `module.x.super_module_tool.config_version = 1`
+
+ - Plus any other best practices we come up with.
+
+## Unresolved Questions
+
+- Do we need a “system”-level layer of config? (i.e. user-level config that is the default for all users of a system,
```

Comment:
-1 on the need for a system-level config.

Reply:
+1 for system level config 

Useful for being able to preset `*.module_defaults.template.*` settings to company defaults.

Reply:
@vinzent Do you have multiple users using the same system for module development or would the "system" level config be provisioned automatically onto users workstations through Puppet, etc.?

Reply:
@scotje we have a central puppet-dev server where we develop puppet modules. nonetheless I would provision the setting by puppet. ;-)

Reply:
@vinzent Thanks for the additional context. Would lack of the system-level config in the initial release cause you to not adopt `pdk config` right away? Trying to narrow down whether that feature is a "must have" or a "nice to have" for the first release.

Reply:
I think we will defer system-level config for now and we can always add it in a subsequent release.